forte/data/data_store.py

```diff
@@ -493,6 +496,134 @@ def get_entry(self, tid: int) -> Tuple[List, int, int]:
         # Otherwise, use ``index_id`` to find the index of the entry.
         raise NotImplementedError
 
+    def co_iterator_annotation_like(
+        self, type_names: List[str]
+    ) -> Iterator[List]:
+        r"""
+        Given two or more type names, iterate their entry lists from beginning to end together.
+
+        For every single type, their entry lists are sorted by the ``begin`` and
+        ``end`` fields. The ``co_iterator_annotation_like`` function will iterate those sorted lists
+        together, and yield each entry in sorted order. This tasks is quite
+        similar to merging several sorted list to one sorted list. We internally
+        use a `MinHeap` to order the order of yielded items, and the ordering
+        is determined by:
+
+            - start index of the entry.
+            - end index of the entry.
+            - the index of the entry type name in input parameter ``type_names``.
+
+        The precedence of those values indicates their priority in the min heap
+        ordering.
+        For example, if two entries have both the same begin and end field,
+        then their order is
+        decided by the order of user input type_name (the type that first
+        appears in the target type list will return first).
+        For entries that have the exact same `begin`, `end` and `type_name`,
+        the order will be determined arbitrarily.
+
+        Args:
+            type_names (List[str]): a list of string type names
+
+        Returns:
+
+            An iterator of entry elements.
+
+        """
+
+        n = len(type_names)
+        # suppose the length of type_names is N and the length of entry list of
+        # one type is M
+        # then the time complexity of using min-heap to iterate
+        # is O(M*log(N))
+
+        # Initialize the first entry of all entry lists
+        # it avoids empty entry lists or non-existant entry list
+        first_entries = []
+        # record the current entry index for elements
+        # pointers[i] is the index of entry at (i)th sorted entry lists
+        # initialize with -1 which means no indexing in the empty list
+        pointers = [-1] * n
+
+        for i, tn in enumerate(type_names):
+            try:
+                if self.__elements[tn]:
+                    first_entries.append(self.__elements[tn][0])
+                    pointers[i] = 0
+            except KeyError as e:  # self.__elements[tn] will be catched here.
+                raise ValueError(
+                    f"Input argument `type_names` to the function contains"
+                    f" a type name [{tn}], which is not recognized."
+                    f" Please input available ones in this DataStore"
+                    f" object: {list(self.__elements.keys())}"
+                ) from e
```

Comment:
We can still keep the `IndexError` even though we know `__elements` should be non-empty, but one more catch here will sometimes help us debug unexpected behaviors